lib/CAPI/Firtool/Firtool.cpp

```diff
@@ -0,0 +1,264 @@
+        return firrtl::PreserveAggregate::OneDimVec;
+      case FIRTOOL_PRESERVE_AGGREGATE_MODE_VEC:
+        return firrtl::PreserveAggregate::Vec;
+      case FIRTOOL_PRESERVE_AGGREGATE_MODE_ALL:
+        return firrtl::PreserveAggregate::All;
+      default: // NOLINT(clang-diagnostic-covered-switch-default)
+        assert(false && "unknown preserve aggregate mode");
+      }
+    },
+    [](firrtl::PreserveAggregate::PreserveMode value) {
+      switch (value) {
+      case firrtl::PreserveAggregate::None:
+        return FIRTOOL_PRESERVE_AGGREGATE_MODE_NONE;
+      case firrtl::PreserveAggregate::OneDimVec:
+        return FIRTOOL_PRESERVE_AGGREGATE_MODE_ONE_DIM_VEC;
+      case firrtl::PreserveAggregate::Vec:
+        return FIRTOOL_PRESERVE_AGGREGATE_MODE_VEC;
+      case firrtl::PreserveAggregate::All:
+        return FIRTOOL_PRESERVE_AGGREGATE_MODE_ALL;
+      default: // NOLINT(clang-diagnostic-covered-switch-default)
+        assert(false && "unknown preserve aggregate mode");
+      }
+    });
+DEFINE_FIRTOOL_OPTION_ENUM(
+    PreserveValues, preserveMode, FirtoolPreserveValuesMode,
+    [](FirtoolPreserveValuesMode value) {
+      switch (value) {
+      case FIRTOOL_PRESERVE_VALUES_MODE_NONE:
+        return firrtl::PreserveValues::None;
+      case FIRTOOL_PRESERVE_VALUES_MODE_NAMED:
+        return firrtl::PreserveValues::Named;
+      case FIRTOOL_PRESERVE_VALUES_MODE_ALL:
+        return firrtl::PreserveValues::All;
+      default: // NOLINT(clang-diagnostic-covered-switch-default)
+        assert(false && "unknown preserve values mode");
+      }
+    },
+    [](firrtl::PreserveValues::PreserveMode value) {
+      switch (value) {
+      case firrtl::PreserveValues::None:
+        return FIRTOOL_PRESERVE_VALUES_MODE_NONE;
+      case firrtl::PreserveValues::Named:
+        return FIRTOOL_PRESERVE_VALUES_MODE_NAMED;
+      case firrtl::PreserveValues::All:
+        return FIRTOOL_PRESERVE_VALUES_MODE_ALL;
+      default: // NOLINT(clang-diagnostic-covered-switch-default)
+        assert(false && "unknown preserve values mode");
+      }
+    });
+DEFINE_FIRTOOL_OPTION_ENUM(
+    BuildMode, buildMode, FirtoolBuildMode,
+    [](FirtoolBuildMode value) {
+      switch (value) {
+      case FIRTOOL_BUILD_MODE_DEBUG:
+        return firtool::FirtoolOptions::BuildModeDebug;
+      case FIRTOOL_BUILD_MODE_RELEASE:
+        return firtool::FirtoolOptions::BuildModeRelease;
+      default: // NOLINT(clang-diagnostic-covered-switch-default)
+        assert(false && "unknown build mode");
+      }
+    },
+    [](firtool::FirtoolOptions::BuildMode value) {
+      switch (value) {
+      case firtool::FirtoolOptions::BuildModeDebug:
+        return FIRTOOL_BUILD_MODE_DEBUG;
+      case firtool::FirtoolOptions::BuildModeRelease:
+        return FIRTOOL_BUILD_MODE_RELEASE;
+      default: // NOLINT(clang-diagnostic-covered-switch-default)
+        assert(false && "unknown build mode");
```

Comment:
nit: instead of `assert(false)`, maybe use `llvm_unreachable`? i'm not sure the recommended pattern to use here, but i generally try to avoid `assert(false)`.

Reply:
This is the comment of `llvm_unreachable`

```
/// Marks that the current location is not supposed to be reachable.
/// In !NDEBUG builds, prints the message and location info to stderr.
/// In NDEBUG builds, if the platform does not support a builtin unreachable
/// then we call an internal LLVM runtime function. Otherwise the behavior is
/// controlled by the CMake flag
///   -DLLVM_UNREACHABLE_OPTIMIZE
/// * When "ON" (default) llvm_unreachable() becomes an optimizer hint
///   that the current location is not supposed to be reachable: the hint
///   turns such code path into undefined behavior.  On compilers that don't
///   support such hints, prints a reduced message instead and aborts the
///   program.
/// * When "OFF", a builtin_trap is emitted instead of an
//    optimizer hint or printing a reduced message.
///
/// Use this instead of assert(0). It conveys intent more clearly, suppresses
/// diagnostics for unreachable code paths, and allows compilers to omit
/// unnecessary code.
```

Note that `llvm_unreachable` sometimes leads to a undefined behavior, I'm not sure if it's suitable for such user input errors, so I replaced it with `assert`.

I'll change it back to `assert` and can discuss this further later.

Reply:
Done.